### PR TITLE
ENT-2506 restore the attachment party signers

### PR DESCRIPTION
**Pull request by tudor-malene**
This restores the Attachment.signers api that was removed by mistake.
It is marked as deprecated and not used.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt [hunk 1]
```diff
@@ -47,10 +48,15 @@ abstract class AbstractAttachment(dataLoader: () -> ByteArray) : Attachment {
...
 
     override fun open(): InputStream = attachmentData.inputStream()
-    override val signers: List<PublicKey> by lazy {
+
+    override val signerKeys: List<PublicKey> by lazy {
         openAsJAR().use(JarSignatureCollector::collectSigners)
     }
 
+    override val signers: List<Party> by lazy {
+        openAsJAR().use(JarSignatureCollector::collectSigningParties)
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt]: Looking at the implementations for `collectSigners` and `collectSigningParties` I'm failing to see why we need to introduce `signerKeys` in the first place. The jar signing API exposes the signers as X509 certs, which has everything you need to convert to a `Party` object.

Can't we just update the doc for `signers` and say that "... even though this returns a list of party objects, it is not required that these parties exist on the network, but rather they are a mapping from the signing key to the X.500 name"

**Comment by shamsasari** [comment]
I don't think we need to introduce `signerKeys`

File: core/src/main/kotlin/net/corda/core/contracts/Attachment.kt [hunk 2]
```diff
@@ -51,11 +52,19 @@ interface Attachment : NamedByHash {
...
+     */
+    @Deprecated("use signerKeys")
+    val signers: List<Party>
+
     /**
      * The keys that have correctly signed the whole attachment.
      * Can be empty, for example non-contract attachments won't be necessarily be signed.
      */
-    val signers: List<PublicKey>
+    val signerKeys: List<PublicKey>
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/contracts/Attachment.kt]: It should have been `DoNotImplement`.

File: core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt (hunk 1 above)
Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt]: I guess we could, but I would find that confusing.
To create inexistent Parties just to hold some keys.

File: core/src/main/kotlin/net/corda/core/contracts/Attachment.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikehearn) [reply core/src/main/kotlin/net/corda/core/contracts/Attachment.kt]: In this case I'll approve adding DoNotImplement to the attachment interface. Risk of anyone having implemented this themselves is extremely low.

File: core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt (hunk 1 above)
Comment (mikehearn) [reply core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt]: It'd be OK to do that even though it's a bit odd, I think. Mostly nobody should care about that API anyway?

Comment (tudor-malene) [reply core/src/main/kotlin/net/corda/core/internal/AbstractAttachment.kt]: We do that currently anyway in the (now) deprecated `signers` method. 

In the future we should use only the `signerKeys` function that does what it says.

File: core/src/main/kotlin/net/corda/core/contracts/Attachment.kt [hunk 3]
```diff
@@ -51,11 +54,20 @@ interface Attachment : NamedByHash {
     @JvmDefault
     fun extractFile(path: String, outputTo: OutputStream) = openAsJAR().use { it.extractFile(path, outputTo) }
 
+    /**
+     * The parties that have correctly signed the whole attachment.
+     * Even though this returns a list of party objects, it is not required that these parties exist on the network, but rather they are a mapping from the signing key to the X.500 name.
+     *
+     * Note: Anyone can sign attachments, not only Corda parties. It's recommended to use [signerKeys].
+     */
+    @Deprecated("use signerKeys")
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/contracts/Attachment.kt]: Capital U, and perhaps a reason why?

File: core/src/main/kotlin/net/corda/core/contracts/Attachment.kt [hunk 4]
```diff
@@ -31,6 +33,7 @@ import java.util.jar.JarInputStream
  */
 @KeepForDJVM
 @CordaSerializable
+@DoNotImplement
```
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/contracts/Attachment.kt]: Add entry on changelog explaining why

**Pull request merged**